### ISSUE TITLE
chap06_Keras_Model_keras_iris_DNN

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#43]
"""
- Tensorflow2.x Keras + iris
- Keras : DNN model 생성을 위한 고수준 API

- 매우 매우 코드가 간결해짐

- y변수 : on ehotencoding
    loss = 'categorical_crossentrophy'
    metrics=['accuracy']
"""

import tensorflow as tf # ver2.x
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import minmax_scale # x변수 전처리
from tensorflow.keras.utils import to_categorical # 원핫코딩과 비슷한것, y변수 전처리
from tensorflow.keras import Sequential # model 생성
from tensorflow.keras.layers import Dense # layer 생성
from tensorflow.keras.models import load_model # model save/load
from sklearn.metrics import accuracy_score

### 1. x, y 공급 data 
iris = load_iris()

- x변수 : 1~4 칼럼
x_data = iris.data
x_data.shape # (150, 4)
x_data = minmax_scale(x_data) # x변수 전처리 : 정규화
- y변수 
y_data = iris.target
y_data.shape # (150, 1)

- reshape(원핫인코딩위해 했던 것 )
-y_data = y_data.reshape(-1, 1)
-y_data.shape # (150, 1)

- [수정]
y_data = to_categorical(y_data) # y변수 전처리 : one hot encoding
y_data.shape # (150, 3)
'''
기존 tensorflow 코드
obj = OneHotEncoder()
y_data = obj.fit_transform(y_data).toarray()
y_data.shape # (150, 3)
'''

- 75 vs 25
x_train, x_val, y_train, y_val = train_test_split(x_data, y_data)
print(x_train.shape) # (112, 4)
print(y_train.shape) # (112, 3)

### 2. keras model 생성
model = Sequential()
model # object info

### 3. model layer
'''
model.add(Dense(node수, input_shape, activation)) # hidden layer1
model.add(Dense(node수, activation)) # hidden layer1 ~ n
'''
- hidden layer1 = [4,12]
model.add(Dense(12, input_shape = (4,), activation = 'relu')) # 1층
- hidden layer2 = [12,6]
model.add(Dense(6, activation = 'relu')) # 2층
- output layer2 = [6,3]
model.add(Dense(3, activation = 'softmax')) # 3층

### 4. model compile : 학습환경 설정
model.compile(optimizer = 'adam', # 최적화 알고리즘(lr 생략)
              loss = 'categorical_crossentropy', # 손실
              metrics=['accuracy']) # 평가 방법

- layer 확인
model.summary()

### 5. model training : train(112) vs val(38)
model.fit(x=x_train, y=y_train,  # 학습용
          epochs=300,
          verbose=1, # 학습과정 횟수
          validation_data=(x_val, y_val) # 평가용
          )
'''
val_loss: 0.1778 - val_accuracy: 0.9474
Epoch 300/300
4/4 [==============================] - 0s 4ms/step - loss: 0.1337 - accuracy: 0.9821 - val_loss: 0.1749 - val_accuracy: 0.9474
'''

### 6. model evaluation : 모델 검증
model.evaluate(x=x_val, y=y_val)
[0.14822138845920563, 0.9473684430122375]

### 7. model save/load
model.save("keras_model_iris.h5")
print("model save")
- 특별하게 경로지정하지 않으면 현재경로에 파일로 저장됨.

new_model = load_model("keras_model_iris.h5")
print(new_model) #object info

### 8. model test : new dataset
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data)

y_pred = new_model.predict(x_test) # 예측치 : 확률값
y_true = y_test # 관측치 : one hot
-비교할 수 없으므로 이것을 맞추어야 한다.

y_pred = tf.argmax(y_pred, axis=1)
y_true = tf.argmax(y_true, axis=1)

from sklearn.metrics import accuracy_score

acc = accuracy_score(y_true, y_pred)
print("accuracy =", acc) # accuracy = 0.9736842105263158
